src/transform/src/union_cancel.rs

```diff
@@ -55,35 +56,83 @@ impl UnionBranchCancellation {
-                        return Some(i);
+                    // Only compare branches with opposite signs
+                    if sign != input_signs[i] {
+                        if let BranchCmp::Inverse = Self::compare_branches(input, &inputs[i]) {
+                            return Some(i);
+                        }
                     }
                 }
                 None
             };
 
-            if let Some(j) = matching_negation(&*base, inputs, 0) {
-                let relation_typ = base.typ();
-                **base = MirRelationExpr::constant(vec![], relation_typ.clone());
-                inputs[j] = MirRelationExpr::constant(vec![], relation_typ);
-            }
+            let base_sign = Self::branch_sign(base);
+            let input_signs = inputs
+                .iter()
+                .map(|input| Self::branch_sign(input))
+                .collect_vec();
 
-            for i in 0..inputs.len() {
-                if let Some(j) = matching_negation(&inputs[i], inputs, i + 1) {
-                    let relation_typ = inputs[i].typ();
-                    inputs[i] = MirRelationExpr::constant(vec![], relation_typ.clone());
+            // Compare branches if there is at least a negated branch
+            if let Some(_) = std::iter::once(&base_sign)
+                .chain(&input_signs)
+                .find(|x| **x == -1)
+            {
+                if let Some(j) = matching_negation(&*base, base_sign, inputs, &input_signs, 0) {
+                    let relation_typ = base.typ();
+                    **base = MirRelationExpr::constant(vec![], relation_typ.clone());
                     inputs[j] = MirRelationExpr::constant(vec![], relation_typ);
                 }
+
+                for i in 0..inputs.len() {
+                    if let Some(j) =
+                        matching_negation(&inputs[i], input_signs[i], inputs, &input_signs, i + 1)
+                    {
+                        let relation_typ = inputs[i].typ();
+                        inputs[i] = MirRelationExpr::constant(vec![], relation_typ.clone());
+                        inputs[j] = MirRelationExpr::constant(vec![], relation_typ);
+                    }
+                }
             }
         }
         Ok(())
     }
 
+    /// Returns the sign of a given union branch. The sign is -1 if the branch contains
+    /// a Negate operator within a chain of Map, Filter and Project operators, and 1
+    /// otherwise.
+    ///
+    /// This sign is pre-computed for all union branches in order to avoid performing
+    /// expensive comparisons of branches with the same sign since they can't possibly
+    /// cancel each other.
+    fn branch_sign(branch: &MirRelationExpr) -> i64 {
+        let mut relation = branch;
+        let mut sign = 1;
+        loop {
+            match relation {
+                MirRelationExpr::Negate { input } => {
+                    sign *= -1;
```

Comment:
You could have sign be a bool. To flip the bool, call `sign ^ true`, a.k.a. sign XOR true.

Reply:
Good point. It's more natural as a bool.